x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -612,6 +622,65 @@ protected Expression maybeSimplifyNegatable(Expression e) {
 
     }
 
+    /**
+     * Look for numerical literals on the right hand side of a {@link BinaryComparison}s that are out of range for the left hand side data
+     * type and replace the comparison by a literal TRUE or FALSE. This rule should be placed after {@link LiteralsOnTheRight}.
+     */
+    private static class OutOfRangeBinaryComparison extends OptimizerRules.OptimizerExpressionRule<BinaryComparison> {
+
+        OutOfRangeBinaryComparison() {
+            super(TransformDirection.DOWN);
```

Comment:
The direction should be up, bottoms-up since the constants are at the bottom.

Reply:
I initially thought this, too; however, semantically the direction doesn't matter for the outcome, and going down has the benefit that we avoid  traversing a whole subtree if we encounter an out of range literal.

E.g. transforming `(a + b*(c-d*(e-f)) < 1E307` is better done top-down because, if we know the data type on the left is int, we don't need to descend into the left hand side at all to know this comparison is always true.